components/local-app/main.go

```diff
@@ -196,23 +186,136 @@ func run(origin, sshConfig string, apiPort int, allowCORSFromPort bool, autoTunn
+			grpcweb.WithCorsForRegisteredEndpointsOnly(false),
+			grpcweb.WithOriginFunc(allowOrigin),
+			grpcweb.WithWebsockets(true),
+			grpcweb.WithWebsocketOriginFunc(func(req *http.Request) bool {
+				origin, err := grpcweb.WebsocketRequestOrigin(req)
+				if err != nil {
+					return false
+				}
+				return allowOrigin(origin)
+			}),
+			grpcweb.WithWebsocketPingInterval(15*time.Second),
+		))
+		if err != nil {
+			logrus.WithError(err).Error("API endpoint failed to start")
+			os.Exit(1)
+		}
+	}()
 	defer grpcServer.Stop()
 	return b.Run()
 }
 
+func connectToServer(loginOpts auth.LoginOpts, reconnectionHandler func()) (*gitpod.APIoverJSONRPC, error) {
+	tkn, err := auth.GetToken(loginOpts.GitpodURL)
+	if err != nil {
+		return nil, err
+	}
+
+	if tkn != "" {
+		// try to connect with existing token
+		client, err := tryConnectToServer(loginOpts.GitpodURL, tkn, reconnectionHandler)
+		if client != nil {
+			return client, err
+		}
+		_, invalid := err.(*ErrInvalidGitpodToken)
+		if !invalid {
+			return nil, err
+		}
+		// existing token is invalid, try again
+		logrus.WithError(err).WithField("origin", loginOpts.GitpodURL).Error()
+	}
+
+	tkn, err = login(loginOpts)
+	if err != nil {
+		return nil, err
+	}
+	return tryConnectToServer(loginOpts.GitpodURL, tkn, reconnectionHandler)
+}
+
+func tryConnectToServer(gitpodUrl string, tkn string, reconnectionHandler func()) (*gitpod.APIoverJSONRPC, error) {
+	wshost := gitpodUrl
+	wshost = strings.ReplaceAll(wshost, "https://", "wss://")
+	wshost = strings.ReplaceAll(wshost, "http://", "ws://")
+	wshost += "/api/v1"
+	client, err := gitpod.ConnectToServer(wshost, gitpod.ConnectToServerOpts{
+		Context:             context.Background(),
+		Token:               tkn,
+		Log:                 logrus.NewEntry(logrus.StandardLogger()),
+		ReconnectionHandler: reconnectionHandler,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	err = validateToken(client, tkn)
+	if err == nil {
+		return client, nil
+	}
+
+	closeErr := client.Close()
+	if closeErr != nil {
+		logrus.WithError(closeErr).WithField("origin", gitpodUrl).Warn("failed to close connection to gitpod server")
+	}
+
+	deleteErr := auth.DeleteToken(gitpodUrl)
+	if deleteErr != nil {
+		logrus.WithError(deleteErr).WithField("origin", gitpodUrl).Warn("failed to delete gitpod token")
+	}
+
+	return nil, err
+}
+
+func login(loginOpts auth.LoginOpts) (string, error) {
+	tkn, err := auth.Login(context.Background(), loginOpts)
+	if tkn != "" {
+		err = auth.SetToken(loginOpts.GitpodURL, tkn)
+		if err != nil {
+			logrus.WithField("origin", loginOpts.GitpodURL).Warnf("could not write token to keyring: %s", err)
+			// Allow to continue
+			err = nil
+		}
+	}
+	return tkn, err
+}
+
+type ErrInvalidGitpodToken struct {
+	cause error
+}
+
+func (e *ErrInvalidGitpodToken) Error() string {
+	return "invalid gitpod token: " + e.cause.Error()
+}
+
+func validateToken(client *gitpod.APIoverJSONRPC, tkn string) error {
```

Comment:
I would have expected this function to be part of the `auth` package - and be tested there (e.g. with a Gitpod client mock)

Reply:
moved to auth